ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/FixedStack.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.vm;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.lang.reflect.Array;
+import java.util.Arrays;
+
+/**
+ * An operand stack for the Ethereum Virtual machine (EVM).
+ *
+ * <p>The operand stack is responsible for storing the current operands that the EVM can execute. It
+ * is assumed to have a fixed size.
+ */
+public class FixedStack<T> {
+
+  private final T[] entries;
+
+  private final int maxSize;
+
+  private int top;
+
+  public static class UnderflowException extends RuntimeException {
+    // Don't create a stack trace since these are not "errors" per say but are using exceptions to
+    // throw rare control flow conditions (EVM stack overflow) that are expected to be seen in
+    // normal
+    // operations.
+    @Override
+    public synchronized Throwable fillInStackTrace() {
+      return this;
+    }
+  }
+
+  public static class OverflowException extends RuntimeException {
+    // Don't create a stack trace since these are not "errors" per say but are using exceptions to
+    // throw rare control flow conditions (EVM stack overflow) that are expected to be seen in
+    // normal
+    // operations.
+    @Override
+    public synchronized Throwable fillInStackTrace() {
+      return this;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public FixedStack(final int maxSize, final Class<T> klass) {
+    checkArgument(maxSize >= 0, "max size must be non-negative");
+
+    this.entries = (T[]) Array.newInstance(klass, maxSize);
+    this.maxSize = maxSize;
+    this.top = -1;
+  }
+
+  public T get(final int offset) {
+    if (offset < 0 || offset >= size()) {
+      throw new IndexOutOfBoundsException();
+    }
+
+    return entries[top - offset];
```

Comment:
shouldn't we check the "top-offset" bounds rather than just offset?

Reply:
offset is always positive.  size() is always top + 1.  So the index is always between zero and top, it's a check that would always pass after asserting those two boundaries on offset.